File: calsnap-web/lib/services/profile-update-service.ts
import { AppConstants } from '@/lib/constants';
import type { BiologicalSex } from '@/lib/models/biological-sex';
import type { ActivityLevel } from '@/lib/models/activity-level';
import type { UserProfile } from '@/lib/models/user-profile';
import type { ProfileDraft } from '@/lib/onboarding/profile-draft';
import { trimmedName } from '@/lib/onboarding/profile-draft';
import {
  ageFromDateOfBirth,
  bmr,
  dailyTarget,
  tdee,
} from '@/lib/nutrition/calculator';

export type MacroKind = 'protein' | 'carbs' | 'fat';

export interface PreviewResult {
  tdee: number;
  dailyTarget: number;
  deficitKcal: number;
  minimumCalories: number;
}

export function preview(input: {
  sex: BiologicalSex;
  dateOfBirth: Date;
  heightCm: number;
  weightKg: number;
  activityLevel: ActivityLevel;
  deficitKcal: number;
}): PreviewResult {
  const age = ageFromDateOfBirth(input.dateOfBirth);
  const bmrValue = bmr(input.weightKg, input.heightCm, age, input.sex);
  const tdeeValue = tdee(bmrValue, input.activityLevel);
  const targetResult = dailyTarget(tdeeValue, input.deficitKcal, input.sex);
  const minimum =
    input.sex === 'male'
      ? AppConstants.Deficit.minCaloriesMale
      : AppConstants.Deficit.minCaloriesFemale;

  return {
    tdee: Math.round(tdeeValue),
    dailyTarget: targetResult.target,
    deficitKcal: targetResult.deficit,
    minimumCalories: minimum,
  };
}

export function apply(
  profile: UserProfile,
  draft: ProfileDraft,
  weightKg: number,
): void {
  const result = preview({
    sex: draft.sex,
    dateOfBirth: draft.dateOfBirth,
    heightCm: draft.heightCm,
    weightKg,
    activityLevel: draft.activityLevel,
    deficitKcal: profile.deficitKcal,
  });

  profile.name = trimmedName(draft);
  profile.sex = draft.sex;
  profile.dateOfBirth = draft.dateOfBirth;
  profile.heightCm = draft.heightCm;
  profile.goalWeightKg = draft.goalWeightKg;
  profile.goalTargetDate = draft.goalTargetDate;
  profile.activityLevel = draft.activityLevel;
  profile.tdee = result.tdee;
  profile.dailyCalorieTarget = result.dailyTarget;
  profile.deficitKcal = result.deficitKcal;
  profile.updatedAt = new Date();
}

export function applyMacroTargets(
  profile: UserProfile,
  proteinPct: number,
  carbsPct: number,
  fatPct: number,
): void {
  profile.macroTargetProteinPct = proteinPct / 100;
  profile.macroTargetCarbsPct = carbsPct / 100;
  profile.macroTargetFatPct = fatPct / 100;
  profile.updatedAt = new Date();
}

export function macroPercentsAreValid(
  protein: number,
  carbs: number,
  fat: number,
): boolean {
  return protein >= 0 && carbs >= 0 && fat >= 0 && protein + carbs + fat === 100;
}

export function normalizedMacroPercents(
  protein: number,
  carbs: number,
  fat: number,
): [number, number, number] {
  if (macroPercentsAreValid(protein, carbs, fat)) {
    return [protein, carbs, fat];
  }
  const sum = protein + carbs + fat;
  if (sum <= 0) {
    return [
      Math.round(AppConstants.Nutrition.defaultMacroProteinPct * 100),
      Math.round(AppConstants.Nutrition.defaultMacroCarbsPct * 100),
      Math.round(AppConstants.Nutrition.defaultMacroFatPct * 100),
    ];
  }
  const p = Math.round((protein / sum) * 100);
  const c = Math.round((carbs / sum) * 100);
  const f = 100 - p - c;
  return [Math.max(0, p), Math.max(0, c), Math.max(0, f)];
}

export function adjustMacroPercents(
  changed: MacroKind,
  newValue: number,
  protein: number,
  carbs: number,
  fat: number,
): [number, number, number] {
  const clamped = Math.min(Math.max(newValue, 0), 100);
  let p = protein;
  let c = carbs;
  let f = fat;

  switch (changed) {
    case 'protein': {
      p = clamped;
      const remaining = 100 - p;
      const otherSum = carbs + fat;
      if (otherSum === 0) {
        c = Math.floor(remaining / 2);
        f = remaining - c;
      } else {
        c = Math.round((remaining * carbs) / otherSum);
        f = remaining - c;
      }
      break;
    }
    case 'carbs': {
      c = clamped;
      const remaining = 100 - c;
      const otherSum = protein + fat;
      if (otherSum === 0) {
        p = Math.floor(remaining / 2);
        f = remaining - p;
      } else {
        p = Math.round((remaining * protein) / otherSum);
        f = remaining - p;
      }
      break;
    }
    case 'fat': {
      f = clamped;
      const remaining = 100 - f;
      const otherSum = protein + carbs;
      if (otherSum === 0) {
        p = Math.floor(remaining / 2);
        c = remaining - p;
      } else {
        p = Math.round((remaining * protein) / otherSum);
        c = remaining - p;
      }
      break;
    }
  }

  return [p, c, f];
}

File: archive/ios/CalSnap/Core/Services/ProfileUpdateService.swift
import Foundation

enum MacroKind: Hashable {
    case protein
    case carbs
    case fat
}

enum ProfileUpdateService {
    struct PreviewResult: Equatable {
        let tdee: Int
        let dailyTarget: Int
        let deficitKcal: Int
        let minimumCalories: Int
    }

    static func preview(
        sex: BiologicalSex,
        dateOfBirth: Date,
        heightCm: Double,
        weightKg: Double,
        activityLevel: ActivityLevel,
        deficitKcal: Int
    ) -> PreviewResult {
        let age = NutritionCalculator.age(from: dateOfBirth)
        let bmr = NutritionCalculator.bmr(
            weightKg: weightKg,
            heightCm: heightCm,
            ageYears: age,
            sex: sex
        )
        let tdeeValue = NutritionCalculator.tdee(bmr: bmr, activityLevel: activityLevel)
        let targetResult = NutritionCalculator.dailyTarget(
            tdee: tdeeValue,
            requestedDeficit: deficitKcal,
            sex: sex
        )
        let minimum = switch sex {
        case .male: AppConstants.Deficit.minCaloriesMale
        case .female: AppConstants.Deficit.minCaloriesFemale
        }
        return PreviewResult(
            tdee: Int(tdeeValue.rounded()),
            dailyTarget: targetResult.target,
            deficitKcal: targetResult.deficit,
            minimumCalories: minimum
        )
    }

    static func apply(
        to profile: UserProfile,
        draft: ProfileDraft,
        weightKg: Double
    ) {
        let result = preview(
            sex: draft.sex,
            dateOfBirth: draft.dateOfBirth,
            heightCm: draft.heightCm,
            weightKg: weightKg,
            activityLevel: draft.activityLevel,
            deficitKcal: profile.deficitKcal
        )

        profile.name = draft.trimmedName
        profile.sex = draft.sex
        profile.dateOfBirth = draft.dateOfBirth
        profile.heightCm = draft.heightCm
        profile.goalWeightKg = draft.goalWeightKg
        profile.goalTargetDate = draft.goalTargetDate
        profile.activityLevel = draft.activityLevel
        profile.tdee = result.tdee
        profile.dailyCalorieTarget = result.dailyTarget
        profile.deficitKcal = result.deficitKcal
        profile.updatedAt = Date.now
    }

    static func applyMacroTargets(
        to profile: UserProfile,
        proteinPct: Int,
        carbsPct: Int,
        fatPct: Int
    ) {
        profile.macroTargetProteinPct = Double(proteinPct) / 100.0
        profile.macroTargetCarbsPct = Double(carbsPct) / 100.0
        profile.macroTargetFatPct = Double(fatPct) / 100.0
        profile.updatedAt = Date.now
    }

    static func macroPercentsAreValid(protein: Int, carbs: Int, fat: Int) -> Bool {
        protein >= 0 && carbs >= 0 && fat >= 0 && protein + carbs + fat == 100
    }

    static func normalizedMacroPercents(protein: Int, carbs: Int, fat: Int) -> (Int, Int, Int) {
        if macroPercentsAreValid(protein: protein, carbs: carbs, fat: fat) {
            return (protein, carbs, fat)
        }
        let sum = protein + carbs + fat
        guard sum > 0 else {
            return (
                Int((AppConstants.Nutrition.defaultMacroProteinPct * 100).rounded()),
                Int((AppConstants.Nutrition.defaultMacroCarbsPct * 100).rounded()),
                Int((AppConstants.Nutrition.defaultMacroFatPct * 100).rounded())
            )
        }
        var p = Int((Double(protein) / Double(sum) * 100).rounded())
        let c = Int((Double(carbs) / Double(sum) * 100).rounded())
        let f = 100 - p - c
        return (max(0, p), max(0, c), max(0, f))
    }

    static func adjustMacroPercents(
        changed: MacroKind,
        newValue: Int,
        protein: Int,
        carbs: Int,
        fat: Int
    ) -> (Int, Int, Int) {
        let clamped = min(max(newValue, 0), 100)
        var p = protein
        var c = carbs
        var f = fat

        switch changed {
        case .protein:
            p = clamped
            let remaining = 100 - p
            let otherSum = carbs + fat
            if otherSum == 0 {
                c = remaining / 2
                f = remaining - c
            } else {
                c = Int((Double(remaining) * Double(carbs) / Double(otherSum)).rounded())
                f = remaining - c
            }
        case .carbs:
            c = clamped
            let remaining = 100 - c
            let otherSum = protein + fat
            if otherSum == 0 {
                p = remaining / 2
                f = remaining - p
            } else {
                p = Int((Double(remaining) * Double(protein) / Double(otherSum)).rounded())
                f = remaining - p
            }
        case .fat:
            f = clamped
            let remaining = 100 - f
            let otherSum = protein + carbs
            if otherSum == 0 {
                p = remaining / 2
                c = remaining - p
            } else {
                p = Int((Double(remaining) * Double(protein) / Double(otherSum)).rounded())
                c = remaining - p
            }
        }

        return (p, c, f)
    }
}
